pkg/controllers/metrics/node/counter/counter.go

```diff
@@ -0,0 +1,166 @@
+	metricTopologyZoneLabel            = "zone"
+	metricOsLabel                      = "os"
+)
+
+var (
+	nodeCountGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: metrics.KarpenterNamespace,
+			Subsystem: "cluster",
+			Name:      "node_count",
+			Help:      "Count of cluster nodes. Broken out by topology and status.",
+		},
+		[]string{
+			metrics.ProvisionerLabel,
+			metricArchLabel,
+			metricConditionDiskPressureLabel,
+			metricConditionMemoryPressureLabel,
+			metricConditionPIDPressureLabel,
+			metricConditionReadyLabel,
+			metricInstanceTypeLabel,
+			metricTopologyRegionLabel,
+			metricTopologyZoneLabel,
+			metricOsLabel,
+		},
+	)
+
+	prometheusLabelsFor = make(map[types.NamespacedName]prometheus.Labels)
+
+	conditionTypeToMetricLabel = map[coreapi.NodeConditionType]string{
+		coreapi.NodeDiskPressure:   metricConditionDiskPressureLabel,
+		coreapi.NodeMemoryPressure: metricConditionMemoryPressureLabel,
+		coreapi.NodePIDPressure:    metricConditionPIDPressureLabel,
+		coreapi.NodeReady:          metricConditionReadyLabel,
+	}
+)
+
+func init() {
+	crmetrics.Registry.MustRegister(nodeCountGaugeVec)
+}
+
+// UpdateCount updates the emitted metric based on the node's current status relative to the
+// past status. If the data for `node` cannot be populated then `nil` should be passed as the
+// argument.
+func UpdateCount(ctx context.Context, name types.NamespacedName, node *coreapi.Node) {
+	currLabels := getLabels(node)
+	pastLabels, isKnown := prometheusLabelsFor[name]
+	switch {
+	case !isKnown && node != nil:
+		handleCreatedNode(ctx, name, currLabels)
+	case isKnown && node == nil:
+		handleDeletedNode(ctx, name, pastLabels)
+	case isKnown && node != nil:
+		handleUpdatedNode(ctx, name, currLabels, pastLabels)
+	default: // An unknown node was deleted.
+	}
+}
+
+func handleCreatedNode(ctx context.Context, name types.NamespacedName, labels prometheus.Labels) {
+	prometheusLabelsFor[name] = labels
+
+	if err := incrementNodeCount(labels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to update count metric for new node [labels=%s]: error=%s", labels, err.Error())
+	}
+}
+
+func handleDeletedNode(ctx context.Context, name types.NamespacedName, labels prometheus.Labels) {
+	delete(prometheusLabelsFor, name)
+
+	if err := decrementNodeCount(labels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to update count metric for deleted node [labels=%s]: error=%s", labels, err.Error())
+	}
+}
+
+func handleUpdatedNode(ctx context.Context, name types.NamespacedName, currLabels prometheus.Labels, pastLabels prometheus.Labels) {
+	// Only report node updates that affect tracked dimensions.
+	if reflect.DeepEqual(currLabels, pastLabels) {
+		return
+	}
+
+	prometheusLabelsFor[name] = currLabels
+
+	if err := decrementNodeCount(pastLabels); err != nil {
+		logging.FromContext(ctx).Warnf("Failed to decrement previous count for updated node [labels=%s]: error=%s", pastLabels, err.Error())
```

Comment:
I was testing this code out and I think there's a false assumption in here with the node properties being set.  It could be that node properties like arch, os, instance type, etc are not set.  

I'm getting floods of this message:

```
karpenter-controller-9f4878f7b-8rt9b manager 2021-09-27T21:54:04.263Z	WARN	controller.NodeMetrics.Count	Failed to decrement previous count for updated node [labels=map[arch: instancetype: os: provisioner:default region: zone:]]: error=inconsistent label cardinality: expected 10 label values but got 6 in prometheus.Labels{"arch":"", "instancetype":"", "os":"", "provisioner":"default", "region":"", "zone":""}	{"commit": "5655c94"}
```